oil_lang/expr_eval.py

```diff
@@ -579,64 +633,243 @@ def _EvalUnary(self, node):
+    # type: (value_t, value_t, Id_t) -> value__Int
+    left_i = self._ValueToInteger(left)
+    right_i = self._ValueToInteger(right)
+
+    if op == Id.Arith_Amp:
+      return value.Int(left_i & right_i)
+    elif op == Id.Arith_Pipe:
+      return value.Int(left_i | right_i)
+    elif op == Id.Arith_Caret:
+      return value.Int(left_i ^ right_i)
+    elif op == Id.Arith_DGreat:
+      return value.Int(left_i >> right_i)
+    elif op == Id.Arith_DLess:
+      return value.Int(left_i << right_i)
+
+    raise NotImplementedError()
+
+  def _ArithLogical(self, left, right, op):
+    # type: (value_t, value_t, Id_t) -> value__Bool
+    UP_left = left
+    UP_right = right
+
+    with tagswitch(left) as lcase:
+      if lcase(value_e.Bool):
+        left = cast(value__Bool, UP_left)
+
+        with tagswitch(right) as rcase:
+          if rcase(value_e.Bool):
+            right = cast(value__Bool, UP_right)
 
-    if node.op.id == Id.Arith_Slash:
-      # NOTE: does not depend on from __future__ import division
+            if op == Id.Expr_And:
+              return value.Bool(left.b and right.b)
+            elif op == Id.Expr_Or:
+              return value.Bool(left.b or right.b)
+
+          else:
+            raise error.InvalidType('Expected Bool', loc.Missing())
+
+      else:
+        raise error.InvalidType('Expected Bool', loc.Missing())
+
+      raise NotImplementedError()
+
+  def _Concat(self, left, right):
+    # type: (value_t, value_t) -> value_t
+    UP_left = left
+    UP_right = right
+
+    with tagswitch(left) as lcase:
+      if lcase(value_e.List):
+        left = cast(value__List, UP_left)
+
+        with tagswitch(right) as rcase:
+          if rcase(value_e.List):
+            right = cast(value__List, UP_right)
+            return value.List(left.items + right.items)
+
+          else:
+            raise error.InvalidType('Expected List', loc.Missing())
+
+      if lcase(value_e.Str):
+        left = cast(value__Str, UP_left)
+        with tagswitch(right) as rcase:
+          if rcase(value_e.Str):
+            right = cast(value__Str, UP_right)
+            return value.Str(left.s + right.s)
+
+          else:
+            raise error.InvalidType('Expected String', loc.Missing())
+
+      if lcase(value_e.MaybeStrArray):
+        left = cast(value__MaybeStrArray, UP_left)
+        with tagswitch(right) as rcase:
+          if rcase(value_e.MaybeStrArray):
+            right = cast(value__MaybeStrArray, UP_right)
+            return value.MaybeStrArray(left.strs + right.strs)
+
+          else:
+            raise error.InvalidType('Expected MaybeStrArray', loc.Missing())
+
+      else:
+        raise error.InvalidType('Expected List or String', loc.Missing())
+
+  def _EvalBinary(self, node):
+    # type: (expr__Binary) -> value_t
+
+    left = _PyObjToValue(self._EvalExpr(node.left))
+    right = _PyObjToValue(self._EvalExpr(node.right))
+
+    if node.op.id == Id.Arith_Plus \
```

Comment:
btw as a shortcut, mycpp supports `node.op.id in (Id.Arith_Plus, IdArith_Minus, ...)`

It expands to basically this chain


----

Another thing we can think about is a 2-level dispatching -- we have those for tokens -- Id and Kind

Reply:
ah great! i'll switch to that in the followup with the rename